grpc/stream.go

```diff
@@ -334,15 +337,13 @@ func (s *stream) closeWithError(err error) error {
 	return s.callErrorListeners(err)
 }
 
-// close changes the stream state to closed and triggers the end event listeners
+// close closes the stream and call end event listeners
+// Note: in the regular closing the io.EOF could come
 func (s *stream) close(err error) {
-	if s.state == closed {
+	if err == nil {
 		return
 	}
 
-	s.logger.WithError(err).Debug("stream is closing")
-
-	s.state = closed
 	close(s.done)
 	s.tq.Queue(func() error {
 		return s.callEventListeners(eventEnd)
```

Comment:
Should this be a separate event for the stream "ending" in one direction but not the other?

Reply:
Sorry, I'm not sure if I'm getting what is the concern here :thinking:

Reply:
Currently this events measn the whole stream has ended - and will in prtactice be called *once*. But now it either will mean "the writing direction is closed /ended"  or will mean "either the writing or reading direction has ended". 

Given that it seems like it will be beneficial to close one or the other and continue getting other/`data` events- it seems beneficial to also be able ot recognise when one or the other direction has ended.

Reply:
>Currently this events measn the whole stream has ended - and will in prtactice be called once. But now it either will mean "the writing direction is closed /ended" or will mean "either the writing or reading direction has ended".

But it's not. The behavior remains. The `stream.on('end', () => {})` will be triggered only once, when the stream is closed (for both writing and reading).

And I see no need for `ending` event since that moment happens exact after calling `stream.end()`

Reply:
I see what you mean, but I think the bigger problem is that I personally was mistaken in my understanding of how this is suppsoed to work.

Having taken some time to look at some nodejs [exampels](https://grpc.io/docs/languages/node/basics/#streaming-rpcs) it seems to me like `.end()` the method *only* closes the *writing* part and the `end` event only signals closing the `reading` part.

Reply:
I have been trying to figure out what happens if you try to write to a stream that that was readonly ... and ... I am not certain what happens as I feel the current examples are not complicated enough :( 

And I have been trying to not try to write completely new examples just to get them wrong :(

Reply:
But it isn't a good idea to look at the nodejs's docs/examples to get an understanding of how our implementation works :shrug: . 

Yes, we have a _kind of_ similar API, but it's not the same and in our docs, we're pretty explicit on how it is supposed to work. Like [`stream.end`](https://k6.io/docs/javascript-api/k6-experimental/grpc/stream/stream-end/) signals that the client finished the sending and [`stream.on('end', () => {})`](https://k6.io/docs/javascript-api/k6-experimental/grpc/stream/stream-on/) happens when the stream closes.

And I believe our current behavior `on`'s end is better in k6's context since it could help our users to put some checks there since we declare that this is the moment when the stream finished and everything that should happen is happen or not :sweat_smile: 

>I have been trying to figure out what happens if you try to write to a stream that that was readonly ... and ... I am not certain what happens as I feel the current examples are not complicated enough :(

I might think that here you mean two different cases:

The first case is about the stream being always read-only (a.k.a server-side stream). In that case, we could learn it from the definitions, I believe, and just return trigger an error (or write a log that there was an attempt to write to read-only stream).

The second case is more interesting :thinking: , but I'm not sure that it is realistic. It's the case when the server finishes the stream while the client still wants to send data. On the other side, I believe even in that case, the last word is on the server to report that it's all done (`io.EOF` in the read from server part).